[#11] 게시글, 댓글 CRUD 구현

hoa0217 opened the pull request:
## 개요
게시글, 댓글 CRUD 및 간단한 타임리프 화면을 만들었습니다.

## 작업사항
- BoardViewController : view return controller
- BoardController, CommentController : rest api controller
- BoardService, CommentService : business logic
- BoardDao, CommentDao : db connection
- boards, board, boardCreate, boardUpdate 화면 생성

## 변경로직
- SpringSecurity : defaultUrl `/boards` (게시판 목록 화면)으로 변경
- MainController 패키지 위치 변경 (domain과 연관 없어보임)

> 단순 crud라 금방 구현할 줄 알았는데 생각보다 코딩할 양이 많더라구요.. 😅 리뷰 부탁드립니다!

hoa0217 committed:
feature: "게시판 엔티티 및 Dto 생성"

hoa0217 committed:
feature: "게시판 Dao 및 Mapper 파일 생성"

hoa0217 committed:
test: "게시판 Dao Test작성"

hoa0217 committed:
refactor: "MainController 패키지 변경"

hoa0217 committed:
fix: "boardID -> board_id 컬럼 오류 수정"

hoa0217 committed:
refactor: "BoardDao 테스트 코드 추가"

hoa0217 committed:
feature: "BoardService 구현 및 테스트 코드 작성"

hoa0217 committed:
style: "마지막 개행 추가"

hoa0217 committed:
refactor: "Response -> BoardInfo로 클래스명 변경"

hoa0217 committed:
feature: "조회수 업데이트 기능 구현 완료"

- Response -> BoardInfo 클래스명 변경
- 조회 시 조회수 업데이트 후 Optional객체로 반환하도록 수정
-

Related: #11

hoa0217 committed:
Merge branch 'develop' into feature/11-board

hoa0217 committed:
feature: "게시판 관련 페이지 생성"

Related: #11

hoa0217 committed:
fix: "@NotBlank Long타입 에러 수정"

- Long타입은 @NotNull을 사용해야합니다.

Related: #11

hoa0217 committed:
fix: "Long타입 비교 에러 수정"

- Long타입은 -127~128 범위를 제외하고 equals로 비교해야합니다.

Related: #11

hoa0217 committed:
refactor: "스프링시큐리티 설정 변경"

- default 페이지 boards로 변경
- 게시물 조회를 제외하고는 인증 필요하도록 requestMatcher 추가

Related: #11

hoa0217 committed:
feature: "Board Controller 개발 완료"

- rest api 권장사항 준수

Related: #11

hoa0217 committed:
refactor: "회원가입 페이지 메인버튼 추가"

Related: #11

hoa0217 committed:
feature: "댓글 도메인 및 Dao 구현"

Related: #11

hoa0217 committed:
feature: "댓글 service, controller 구현"

Related: #11

hoa0217 committed:
refactor: "댓글 기능 추가에 따른 게시판 리팩토링"

- 게시글 상세 조회 시 댓글도 함께 조회한다.
- 게시글 삭제 히 댓글도 함께 삭제한다.

Related: #11

hoa0217 committed:
refactor: "댓글 기능 추가에 따른 게시판 화면 수정"

- 게시글 상세 화면에 댓글 CRUD 추가

Related: #11

hoa0217 committed:
refactor: "스프링 시큐리티 수정"

- 게시판 조회 화면을 제외하면 모든 요청은 인증이 필요하다.

Related: #11

hoa0217 committed:
refactor: "테스트코드 리팩토링"

Related: #11

hoa0217 committed:
style: "Dto inner 클래스 명시"

Related: #11

hoa0217 committed:
refactor: "사용하지 않는 userDetails 제거"

Related: #11

hoa0217 committed:
refactor: "Update 페이지 버튼 변경"

- 목록으로 ➡️ 취소

Related: #11

hoa0217 committed:
style: "단어 변경"

- 게시판 ➡️ 게시글

Related: #11

f-lab-fi commented on src/main/java/com/smart/board/controller/BoardViewController.java:
```diff
@@ -0,0 +1,51 @@
+    model.addAttribute("board", boardService.getBoardByBoardId(boardId));
+    return "boardUpdate";
+  }
+}
```
Board에 대한 API와 View(thymeleaf)를 제공하기 위한 코드를 이렇게 나누셨는데요, 혹시 thymeleaf를 사용하지 않고, HTML 페이지만 만들고 API를 RESTful하게 만들어보실 수 있을까요?

f-lab-fi commented on src/main/java/com/smart/comment/controller/dto/CommentDto.java:
```diff
@@ -0,0 +1,102 @@
+  }
+
+  @Getter
+  @NoArgsConstructor(access = AccessLevel.PROTECTED)
+  @AllArgsConstructor
+  @Builder
+  public static class UpdateRequest {
+
+    @NotNull
+    private Long commentId;
+
+    @NotBlank
+    private String content;
+
+    @NotNull
+    private Long userId;
+
+    public Comment toEntity() {
+      return Comment.builder()
+          .commentId(getCommentId())
+          .content(getContent())
+          .updateDate(LocalDateTime.now())
+          .userId(getUserId())
+          .build();
+    }
+  }
+
+  @Getter
+  @NoArgsConstructor(access = AccessLevel.PROTECTED)
+  @AllArgsConstructor
+  @Builder
+  public static class DeleteRequest {
+
+    @NotNull
+    private Long commentId;
+
+    @NotNull
+    private Long userId;
+  }
+}
```
이렇게 CommentDto 안에 모든 클래스가 static으로 존재한다면, 차라리 별도의 클래스를 만드는게 좋지 않을까요~?
CommentDto 안에 존재하여서 얻을 수 있는 이점이 있을까요?

hoa0217 replied:
별도의 클래스로 만들면, 너무 많은 DTO클래스가 생겨 나중에 관리가 힘들 것 같다는 생각이 들었습니다.
또한 하나의 도메인 DTO에 inner클래스로 묶음으로서 응집력도 높일 수 있을거라 생각했습니다!

f-lab-fi commented on src/main/java/com/smart/global/error/NotFoundBoardException.java:
```diff
@@ -0,0 +1,10 @@
+package com.smart.global.error;
+
+public class NotFoundBoardException extends RuntimeException {
+
+  private static final String MESSAGE = "해당 게시물을 찾을 수 없습니다.";
+
+  public NotFoundBoardException() {
+    super(MESSAGE);
+  }
+}
```
NotFound**Comment**Exception에 대해서도 예외를 별도로 정의해줬는데요, 이런식으로 모든 엔티티(Entity)마다 찾지 못했다는 예외를 만들어주다보면 너무 많은 예외를 만들어야 할 수도 있습니다.
이런 문제를 피하려면 어떻게 하면 좋을까요?

hoa0217 replied:
NotFoundEntityException을 하나만 만들고, 예외를 던질 때 마다 메시지에 어떤 엔티티인지 명시해주는 것도 좋은 방법일거같습니다!

f-lab-fi commented on src/main/java/com/smart/board/controller/dto/BoardDto.java:
```diff
@@ -0,0 +1,122 @@
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
+public class BoardDto {
+
```
https://github.com/f-lab-edu/s-mart/pull/17/files#r1170183554

위 리뷰와동일합니다. :)

f-lab-fi replied:
그러고보니 Board라는 이름이 좀 어색해 보여요.
Board는 보통 '게시판'이라는 의미로 사용되는데, '게시글'과는 분명 다른 의미의 단어라고 생각됩니다.
게시글이라면 Post나 Article 같은 단어가 더 자연스러울 것 같습니다. :)

f-lab-fi commented on src/main/java/com/smart/board/domain/Board.java:
```diff
@@ -0,0 +1,23 @@
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@Builder
+@AllArgsConstructor
+public class Board {
+
```
그러고보니 Board라는 이름이 좀 어색해 보여요.
Board는 보통 '게시판'이라는 의미로 사용되는데, '게시글'과는 분명 다른 의미의 단어라고 생각됩니다.
게시글이라면 Post나 Article 같은 단어가 더 자연스러울 것 같습니다. :)

hoa0217 replied:
맞습니다 ㅋㅋ 게시판과 게시글.. 헷갈렸어요.. 수정하겠습니다!

f-lab-fi commented on src/test/java/com/smart/board/dao/BoardDaoTest.java:
```diff
@@ -0,0 +1,137 @@
+    assertThat(boardInfo.getViewCount()).isEqualTo(board.getViewCount()+1);
+  }
+
+  @Test
+  @DisplayName("생성하지 않은 게시글 업데이트하면 당연히 해당 게시물을 찾을 수 없다.")
+  public void updateBoard_NotExistingBoard() {
+    Board updateBoard = BoardDto.UpdateRequest.builder()
+        .boardId(-1L)
+        .title("update title")
+        .content("update content")
+        .userId(board.getUserId())
+        .build()
+        .toEntity();
+
+    boardDao.updateBoard(updateBoard);
+
+    Optional<BoardDto.BoardInfo> optionalBoardInfo = boardDao.getBoardByBoardId(-1L);
+    assertThat(optionalBoardInfo.isEmpty()).isTrue();
+  }
```
이건 그러고보니 존재하지 않는 게시글을 업데이트 하려고하면 boardId에 해당하는 게시글을 찾을 수 없는거니, NotFoundBoardException이 테스트 메서드로 던져지도록 해야하지 않을까요? 테스트 메서드에서는 그걸 검증하도록 하구요.

f-lab-fi commented on src/test/java/com/smart/board/service/BoardServiceTest.java:
```diff
@@ -0,0 +1,223 @@
+  BoardService boardService;
+
+  BoardDto.BoardInfo boardInfo;
+
+  BoardDto.UpdateRequest updateRequest;
+
+  BoardDto.DeleteRequest deleteRequest;
+
+  @BeforeEach
+  public void inject_Mock() {
+    boardDao = mock(BoardDao.class);
+    commentDao = mock(CommentDao.class);
+    boardService = new BoardService(boardDao, commentDao);
+  }
+
+  @BeforeEach
+  public void create_Test_Data() {
+    boardInfo = BoardDto.BoardInfo.builder()
+        .boardId(1L)
+        .title("title")
+        .content("content")
+        .viewCount(1L)
+        .userId(1L)
+        .nickname("nickname")
+        .build();
+
+    updateRequest = BoardDto.UpdateRequest
+        .builder()
+        .boardId(1L)
+        .title("update title")
+        .content("update content")
+        .userId(1L)
+        .build();
+
+    deleteRequest = BoardDto.DeleteRequest
+        .builder()
+        .boardId(1L)
+        .userId(1L)
+        .build();
+  }
```
```
  BoardDto.BoardInfo boardInfo;

  BoardDto.UpdateRequest updateRequest;

  BoardDto.DeleteRequest deleteRequest;
```

위 3가지를 테스트 클래스에서 공유하는 경우 아래 있는 테스트 코드들이 병렬적으로 실행될 수 있을까요?
그리고, 테스트 코드가 병렬적으로 실행되는게 중요한 이유는 무엇일까요?

f-lab-fi replied:
현재 테스트 인스턴스 내에서 해당 필드들이 공유되면서, 각각이 BeforeEach에 의해 매 테스트 메서드 실행 전 초기화 되고 있어서 여쭤보는 내용입니다. :)

hoa0217 replied:
- 테스트 클래스는 `@Test` 메서드 1개 당 인스턴스를 새로 생성하여 테스트를 실행합니다. 따라서 각자 다른 인스턴스이기 때문에 인스턴스 변수를 공유하지 않고 독립적으로 병렬적으로 실행할 수 있습니다.

- 이게 중요한 이유는 테스트 메서드는 서로가 서로에게 영향을 주거나 받으면 안되기 때문입니다. 만약 종속성이 있다면 메서드 실행 순서에 따라 결과가 달라짐으로 순차적으로 실행해야 합니다. 또한 병렬적으로 실행함으로써 테스트 시간을 단축할 수 있습니다.

> 위에 필드로 따로 빼고 `@BeforeEach`에서 초기화하는 이유는 테스트 메서드 내부에 중복해서 코드를 작성하기보다는 Junit의 기능을 활용하는게 더 깔끔하고 효율적이라고 판단했습니다.

잘못된 내용이 있다면 피드백 부탁드립니다!

f-lab-fi replied:
@hoa0217 

넵 테스트 인스턴스는 기본값이 메서드마다 생성되는게 맞습니다. (@TestInstance 관련)
시간되실때 한번 병렬적으로 실행되도록 만들어보시죠. :)

hoa0217 committed:
refactor: "도메인명 변경, DTO 분리, 예외 통합"

- Board ➡️ Post
- Dto inner class 분리
- NotFoundEntityException 으로 통합

Related: #11

hoa0217 committed:
refactor: "api url 변경"

Related: #11

hoa0217 committed:
refactor: "명칭 변경"

Related: #11

hoa0217 committed:
refactor: "dao ➡️ repository"

Related: #11

hoa0217 committed:
fix: "오타수정"

Related: #11

hoa0217 committed:
refactor: "테스트 코드 명칭 수정"

Related: #11

f-lab-fi commented on src/main/java/com/smart/board/controller/BoardViewController.java:
```diff
@@ -13,27 +12,25 @@
+    BoardDetailDto boardDetail = boardService.getBoardDetailByPostId(postId);
+    model.addAttribute("post", boardDetail.getPostDto());
+    model.addAttribute("comments", boardDetail.getCommentDtos());
     model.addAttribute("userDetails", userDetails);
```
Board -> Post로 바뀌었는데, BoardDetail은 따로 존재하는 개념인건가요?
이름이 혼재되어 있어서 어떤 의미인지 파악이 어려워보입니다. :)

f-lab-fi commented on src/main/java/com/smart/board/controller/dto/BoardDetailDto.java:
```diff
@@ -0,0 +1,21 @@
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@AllArgsConstructor
+@Builder
+public class BoardDetailDto {
+
+  PostReadDto postDto;
+
+  List<CommentReadDto> commentDtos;
+}
```
이 DTO가 게시글 보기 페이지에서 보이는 내용 같네요.
Board라는 네이밍이 남아있어서 Board, Post가 어떤역할을 하는지 혼동되게 하는 것 같습니다.

아래 리뷰와 같은 맥락입니다. 

https://github.com/f-lab-edu/s-mart/pull/17/files/06fd09562fa79ccf41d63d3607bd0b46878e6a90..2bff06c1861453ed838fcaad20cfa939f218d4db#r1176818813

f-lab-fi commented on src/main/java/com/smart/board/service/BoardService.java:
```diff
@@ -14,67 +23,98 @@
+    commentRepository.update(updateDto.toEntity());
+  }
+
+  @Transactional
+  public void deleteComment(CommentDeleteDto deleteDto, Long loginUserId) {
+    checkPermission(deleteDto.getUserId(), loginUserId);
+    checkExistComment(deleteDto.getCommentId());
+    commentRepository.deleteByCommentId(deleteDto.getCommentId());
+  }
```
CommentDeleteDto에 해당 댓글의 작성자를 보내서 검증하는 과정을 거치고 있는데요, 이렇게 하지말고 commentRepository에 Comment 하나를 조회하는 API를 만들어서 Comment를 먼저 조회한 후, Comment에 들어있는 작성자와 loginUserId를 비교해주면 어떨까요? 이렇게하면 CommentDeleteDto에 해당 댓글의 작성자가 보내질 필요가 없어보입니다.

f-lab-fi replied:
심지어 이건 보안상으로도 문제가 있어보입니다.
클라이언트가 댓글의 userId를 본인의 userId로 바꿔서 보내는 경우 남의 댓글을 삭제할 수 있지 않나요?

f-lab-fi replied:
1. 댓글 먼저 조회 -> 조회를 실패하는 경우 -> NotFoundEntityException -> @RestControllerAdvice에서 404
2. 댓글의 작성자 == loginUserId
3. 같으면 댓글 삭제처리 -> 200 OK
4. 다르면 예외 -> PermissionDeniedException -> @RestControllerAdvice에서 403

401 -> 인증이 되지 않았음.
403 -> (인증은 된 상태라고 가정) 허가되지 않은 액션

f-lab-fi commented on src/main/java/com/smart/board/repository/CommentRepository.java:
```diff
@@ -0,0 +1,25 @@
+import java.util.Optional;
+import org.apache.ibatis.annotations.Mapper;
+
+@Mapper
+public interface CommentRepository {
+
+  List<CommentReadDto> findByPostId(Long postId);
+
+  void deleteByPostId(Long postId);
+
+  Optional<CommentReadDto> findByCommentId(Long commentId);
```
Repository는 가급적 CommentReadDto 같은 표현 계층에 존재하는 DTO에 의존하지 않아야합니다.
Comment들의 List를 조회해온 후 Service에서 CommentReadDto로 변환하는게 재사용성이 높은 Repository가 됩니다.

hoa0217 replied:
넵 수정했습니다!

f-lab-fi commented on src/main/java/com/smart/board/controller/dto/comment/CommentDeleteDto.java:
```diff
@@ -0,0 +1,21 @@
+  @NotNull
+  private Long commentId;
+
+  @NotNull
+  private Long userId;
```
https://github.com/f-lab-edu/s-mart/pull/17#discussion_r1176836920

위 리뷰와 관련된 내용입니다. (_ _)

f-lab-fi commented on src/main/java/com/smart/board/controller/dto/comment/CommentUpdateDto.java:
```diff
@@ -0,0 +1,36 @@
+  private String content;
+
+  @NotNull
+  private Long userId;
```
https://github.com/f-lab-edu/s-mart/pull/17#discussion_r1176836920

마찬가지입니다.

f-lab-fi commented on src/main/java/com/smart/board/controller/dto/post/PostDeleteDto.java:
```diff
@@ -0,0 +1,22 @@
+  @NotNull
+  private Long postId;
+
+  @NotNull
+  private Long userId;
```
https://github.com/f-lab-edu/s-mart/pull/17#discussion_r1176836920

마찬가지!

f-lab-fi commented on src/main/java/com/smart/board/repository/CommentRepository.java:
```diff
@@ -0,0 +1,25 @@
+import java.util.Optional;
+import org.apache.ibatis.annotations.Mapper;
+
+@Mapper
+public interface CommentRepository {
```
List로 구현된 CommentRepository가 있다고 생각하고 인터페이스를 만들어보면 JPA나 기타 저장기술들과 동일한 인터페이스를 맞출 수 있음

hoa0217 committed:
refactor: "DB의존성제거 및 Collection사용"

Related: #11

hoa0217 committed:
refactor: "api 분리 및 service에서 의존하는 레파지토리 변경"

Related: #11

hoa0217 committed:
refactor: "DB의존성제거 및 Collection 사용 (User)"

Related: #11

hoa0217 committed:
refactor: "화면단 api 수정"

Related: #11

hoa0217 committed:
fix: "model attribute 이름 수정"

Related: #11

hoa0217 committed:
fix: "transactional 어노테이션 추가"

Related: #11

hoa0217 committed:
fix: "조회수 증가 로직 추가"

Related: #11

hoa0217 committed:
refactor: "중복메서드 리팩토링"

Related: #11

hoa0217 committed:
style: "import문 정리"

Related: #11

hoa0217 committed:
chore: "develop merge"

Related: #17

hoa0217 committed:
refactor: "조회 로직 변경"

- 본인글은 조회수를 증가하지 않도록 로직 변경

Related: #11

hoa0217 committed:
test: "조회 로직 변경"

Related: #11

f-lab-fi commented on src/main/java/com/smart/board/service/BoardService.java:
```diff
@@ -25,60 +25,74 @@ public class BoardService {
-        .postDtos(postDtos)
-        .build();
+  public List<PostReadDto> getPosts() {
+    List<PostReadDto> postReadDtos = new ArrayList<>();
+    for (Post post : postRepository.findAll()) {
+      User user = findUserByUserId(post.getUserId());
+      postReadDtos.add(PostReadDto.toDto(post, user));
+    }
+    return postReadDtos;
```
이 부분은 Stream의 map을 활용하는게 더 가독성 높은 코드가 됐을 것 같습니다!

그리고 `User user = findUserByUserId(post.getUserId());` 로 User는 조회해줄 필요가 없어보입니다. Post나 PostReadDto나 모두 User가 아니라, Long형인 userId 만을 사용하고 있는데요, 이러면 User를 조회해줄 필요 없이 PostReadDto.toDto에서 파라미터로 Post만 전달해줬어도 userId를 얻는데는 문제가 없어보입니다.

hoa0217 replied:
- 넵 stream으로 변경해보겠습니다.
- 음.. user를 조회하는 이유는 프론트 화면에서 user의 nickname 필드가 필요하기 때문에, 쿼리에서 merge해서 가져오는 대신 service단에서 조합을 해주는 과정입니다. 아마 저희가 redis를 쓴다면 프론트에 userId만 던져주고 알아서 redis에서 user정보를 가져오면 될거같은데.. 현재는 구현을 못한터라 user조회가 필요하다고 생각합니다.

f-lab-fi replied:
@hoa0217 

아.. nickname도 조회해주고 있군요. ㅎㅎㅎㅎ
그런거라면 조회해줘도 괜찮을 것 같습니다. nickname이 있는지 못봤어요 ㅠ

f-lab-fi replied:
요건 프론트에서 조회하는게 더 부자연스러울 것 같아요~

f-lab-fi commented on src/main/java/com/smart/board/service/BoardService.java:
```diff
@@ -25,60 +25,74 @@ public class BoardService {
+      post.updateViewCount();
+      postRepository.update(post);
+    }
   }
```
이 부분 코드에서 리뷰 드릴만한 내용이 몇가지 있습니다.
다만, 모두 하나의 주제를 다루는 내용입니다.

**첫번째로,** **CustomUserDetails**는 오직 getUserId를 위해서만 존재하고 있습니다. 이러면 CustomUserDetails를 통째로 넘길 필요가 있을까요?

**두번째로**, CustomUserDetails가 domain 패키지로 들어가 있는 것 같은데, 제가 볼 때 CustomUserDetails가 과연 domain 패키지에 있는게 적절한지 의심됩니다. 해당 클래스는 아마 시큐리티 관련해서 유저 정보를 담는 클래스 아닌가요? 그렇다면 이것은 표현 계층(정화님 코드에서는 controller)에 존재하는게 적절해보이는 클래스입니다.

**세번째로**, 해당 글의 작성자와 현재 로그인한 유저가 같지 않아야만 viewCount 를 올려준다는 것은 우리가 하나의 '업무 규칙'이라고 부를만한 것입니다. 즉, 비즈니스 로직이라고 부를만한겁니다. 이런 로직은 가급적 컨트롤러나 서비스가 아니라 도메인에 존재하는게 좋습니다. 그래야 도메인 객체를 봤을 때 업무 규칙을 빠르게 파악할 수 있고, 도메인 객체가 응집력 있는 코드가 되거든요.

어떻게 바꿔야할지 잘 감이 안오실 수도 있어요. 이 부분은 멘토링 시간에 이야기 해볼테니 고민해보시고 잘 모르겠다면 정화님 생각을 우선 정리해보시면 좋을 것 같습니다. :)

hoa0217 replied:
1. CustomUserDetails은 로그인 하지 않은 상태에서는 null 값입니다. 이 null에 대한 체크(로그인상태인지 비로그인상태인지) 및 분기처리를 컨트롤러단에서 하기보단 서비스에서 하는게 맞다 생각하여 서비스로 넘겼습니다.
2. 넵 맞는 말씀이십니다. 표현계층에 존재하는게 적절합니다.
3. validation과 같은 비즈니스로직이 post의 updateViewCount메서드로 가는게 맞겠네요. 수정해보겠습니다!

f-lab-fi replied:
@hoa0217 

글쿤요. 그런데, 로그인한 상태인지 아닌지는 다분히 표현 계층에 의존적인 정보입니다.
서비스를 표현 계층에 의존적이지 않도록 만들기 위해 고려해봐야 할 것은, 표현 계층이 HTTP가 아니라 다른걸로 바뀔 수 있다는걸 가정해봐야합니다.

표현 계층은 지금 처럼 HTTP가 될 수도 있지만, 때론 MQ나 Kafka에서 컨슘한 메시지가 될 수도 있고, 어디선가 발행된 이벤트에 대한 리스너가 될 수도 있습니다.
MQ나 Kafka 혹은 이벤트 역시 로그인된 상태인지 아닌지를 구분할 필요가 있을지 생각해봐야합니다.
**보통 저런 메시지는 로그인된 상태를 구분하지 않으며, 설령 뭔가 구분이 필요하더라도 그 구분을 null로 해주는 것은 좋지 않습니다.** null 보다는 명시적으로 빈 값을 처리할 수 있는 방법을 고려해야합니다.

hoa0217 replied:
멘토님!! 고민하다가 방안을 생각하였는데, CustomUserDetails 객체를 표현계층에서 인자로 가져오는게아니라, 애초에 Long타입인 userId를 인자로 가져오면 굳이 null체크를 하지 않고 바로 Id를 넘길 수 있더라구요!! 일단 비로그인한 사람은 id가 null값인게 맞는거같아 이를 default로 가져오게끔하는 어노테이션을 새로 생성하였습니다!!

```java
package com.smart.global.annotation;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import org.springframework.security.core.annotation.AuthenticationPrincipal;

@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.PARAMETER)
@AuthenticationPrincipal(expression = "#this == 'anonymousUser' ? null : #this.getUserId()")
public @interface AuthUserId {

}

```
컨트롤러
```java
  @GetMapping("/post/{postId}")
  public PostReadDto getPostByPostId(@PathVariable Long postId, @AuthUserId Long userId) {
    return boardService.getPostByPostId(postId, userId);
  }
```
서비스
```java
  @Transactional
  public PostReadDto getPostByPostId(Long postId, Long loginId) {
    Post post = findPostByPostId(postId);
    post.updateViewCount(loginId);
    postRepository.save(post);
    User user = findUserByUserId(post.getUserId());

    return PostReadDto.toDto(post, user);
  }
```

hoa0217 replied:
아 그리고 궁금한점 2가지가 있습니다.
1. `보통 저런 메시지는 로그인된 상태를 구분하지 않으며, 설령 뭔가 구분이 필요하더라도 그 구분을 null로 해주는 것은 좋지 않습니다.`
  - 조회로직은 전에 슬랙에서 질문드렸었는데,, 로그인한 회원ID와 게시글 소유ID를 비교하여 조회수를 증가시킨다는 비즈니스로직으로 구현하기로 결정했기 때문에 로그인유무의 구분은 필요하다고 생각합니다,, 기술적인 이유때문에 비즈니스로직을 또 바꿔야하는게 맞을까요?
2. `null 보다는 명시적으로 빈 값을 처리할 수 있는 방법을 고려해야합니다.` 라고 하시면 대부분 어떤 값으로 구분하나요??

f-lab-fi commented on src/main/java/com/smart/user/repository/UserRepositoryImpl.java:
```diff
@@ -0,0 +1,27 @@
+@Repository
+public class UserRepositoryImpl implements UserRepository {
+
+  private Map<Long, User> users = new HashMap<>();
+  private static Long sequence = 0L;
+
+  @Override
+  public Optional<User> findByUserId(Long userId) {
+    return Optional.ofNullable(users.get(userId));
+  }
+
+  @Override
+  public Long save(User user) {
+    user.setUserId(sequence++);
+    users.put(user.getUserId(), user);
+    return user.getUserId();
+  }
```
이러면 sequence가 0인 User부터 생성되지 않나요~?
다른 레포지토리들도 마찬가지 인 것 같습니다.

hoa0217 replied:
헉..... 아 보통 양수로 시작하나보네요 수정하겠습니다.

hoa0217 committed:
refactor: "repository update메서드 제거"

Related: #11

hoa0217 committed:
refactor: "sequence 시작 값 1로 변경"

hoa0217 committed:
fix: "update 에러 수정"

Related: #11

hoa0217 committed:
refactor: "로그인 체크 controller에게 위임 + 도메인에게 조회수증가 비즈니스로직 위임"

Related: #11

hoa0217 committed:
feature: "로그인한 userId 가져오는 어노테이션 생성"

hoa0217 committed:
refactor: "@AuthUserId 어노테이션 적용"

hoa0217 committed:
test: "조회수증가 테스트 비로그인 case 추가"

f-lab-fi commented on src/main/java/com/smart/board/domain/Post.java:
```diff
@@ -25,7 +25,9 @@ public void setPostId(Long postId) {
+  public void updateViewCount(Long userId) {
+    if (!this.userId.equals(userId)) {
+      viewCount++;
+    }
```
👍

hoa0217 committed:
refactor: "for문 스트림으로 변경"

Related: #11

hoa0217 committed:
refactor: "uerId ➡️ loginUserId"

Related: #11
